### ISSUE TITLE
string coordinate gets converted to object coordinate upon addition of variable to dataset

### ISSUE THREAD

**Issue by anntzer** [pydata/xarray#627]
With the current HEAD, consider

```
import numpy as np
from xray import *

ds = Dataset({"1": DataArray(np.zeros(3), dims=["a"], coords={"a": list("xyz")})})
print(ds)
ds["2"] = DataArray(np.zeros(2), dims=["a"], coords={"a": list("xy")})
print(ds)
```

This outputs

```
<xray.Dataset>
Dimensions:  (a: 3)
Coordinates:
  * a        (a) <U1 'x' 'y' 'z'
Data variables:
    1        (a) float64 0.0 0.0 0.0
<xray.Dataset>
Dimensions:  (a: 3)
Coordinates:
  * a        (a) object 'x' 'y' 'z'
Data variables:
    1        (a) float64 0.0 0.0 0.0
    2        (a) float64 0.0 0.0 nan
```

Note that the dtype of the `a` coordinate got changed after the assignment.

Python3.5, numpy 1.10.1, xray master (6ea7eb2b388075cc838c5ddf0ddaa47020cfcb89)

With 0.6.0 the coordinate is of object dtype both before and after.  I forgot why I tried master but I must have had a good reason...

**Comment by shoyer** [comment]
Thanks for the report -- this does seem like a bug to me. We do end up converting from fixed width strings to object quite often (to get around limitations of NumPy's strings), but adding in a variable should not change the type of any existing variables.

It's possible the best fix here will be to convert all string arrays into an object type as soon as they are loaded into xray objects.

**Comment by anntzer** [comment]
object arrays look fine to me, but I don't have a strong opinion either.

**Comment by anntzer** [comment]
The bug is still valid.

**Comment by max-sixty** [comment]
Given that this is an index, are we dependent on pandas here?

**Comment by max-sixty** [comment]
If this isn't an upstream pandas issue, let's reopen 

(I'm not sure why @stalebot closed this - looks like it's not respecting the comments?)

**Comment by anntzer** [comment]
I don't know whether it's pandas or xarray's fault, but the behavior is definitely still present.

**Comment by andersy005** [comment]
This issue appears to have been resolved: 

```python
In [1]: import xarray as xr
 
In [2]: import numpy as np

In [3]: ds = xr.Dataset({"1": xr.DataArray(np.zeros(3), dims=["a"], coords={"a": list("xyz")})})

In [4]: ds
Out[4]: 
<xarray.Dataset>
Dimensions:  (a: 3)
Coordinates:
  * a        (a) <U1 'x' 'y' 'z'
Data variables:
    1        (a) float64 0.0 0.0 0.0

In [5]: ds["2"] = xr.DataArray(np.zeros(2), dims=["a"], coords={"a": list("xy")})

In [6]: ds
Out[6]: 
<xarray.Dataset>
Dimensions:  (a: 3)
Coordinates:
  * a        (a) <U1 'x' 'y' 'z'
Data variables:
    1        (a) float64 0.0 0.0 0.0
    2        (a) float64 0.0 0.0 nan

```

I tested this with

```python
xarray: 0.17.0
pandas: 1.2.3
numpy: 1.20.1
```

Should we close this?

**Comment by anntzer** [comment]
I have not used xarray recently, so I have no opinion there.

**Comment by keewis** [comment]
it seems this should have been closed by #4759 (bisecting reveals that as the first good commit)
